### PR TITLE
Add a simple utility for debugging that gives supended task backtraces

### PR DESCRIPTION
**Pull request by biddisco**
This should help @parsa with python dumps of stacktraces

### CODE REVIEW

File: hpx/util/debug/thread_stacktrace.hpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    namespace detail {
+        typedef hpx::lcos::local::spinlock mutex_type;
+        typedef std::lock_guard<mutex_type> scoped_lock;
+        //
+        static mutex_type debug_mutex;        
+    }
+    
+    // ------------------------------------------------------------------------
+    // return a vector of suspended/other task Ids
+    static std::vector<hpx::threads::thread_id_type> get_task_ids(
```
Comment (hkaiser) [inline hpx/util/debug/thread_stacktrace.hpp]: Why is this function static?

**Comment by hkaiser** [comment]
We discussed this yesterday and here is a summary of what's required:

- all implementation should be moved to a cpp file (otherwise the code may not be included into the HPX core binary)
- all functions should be `HPX_EXPORT`ed to force the linker to include them into the binary and to make the functions accessible
- the functions shouldn't be `static`
- the locking is not needed as the internal data structures are already protected properly (see https://github.com/STEllAR-GROUP/hpx/blob/master/hpx/runtime/threads/policies/thread_queue.hpp#L936)

**Comment by biddisco** [comment]
I've addressed your comments. Please take another look.

**Comment by biddisco** [comment]
I just (hopefully) fixed the inspect fails. I don't like the fact that it wants #include <string> in both the header and the cpp ... when one includes the other

**Comment by hkaiser** [comment]
> I don't like the fact that it wants #include in both the header and the cpp ... when one includes the other

Well, if somebody includes the header file, and it's the only one referring to `std::string` (or similar), it clearly needs to `#include <string>`. For the cpp file, I can't think of a way to reliably tell inspect which includes to require and which not. Do you have any suggestion?

**Comment by biddisco** [comment]
I do not. Unless we can get inspect to open all the #includes that are included by a file and then ignore required #includes that are inside the included files.

**Comment by K-ballo** [comment]
> I do not. Unless we can get inspect to open all the #includes that are included by a file and then ignore required #includes that are inside the included files.

That would go against the very nature of the inspect check.

**Comment by biddisco** [comment]
I thought someone would say that.

**Pull request merged**